### PR TITLE
Implement Async S3PrefixSensor

### PR DESCRIPTION
**Pull request by phanikumv**
Add Operator, Hook and Trigger for Async S3PrefixSensor
Part of #66

### CODE REVIEW

File: astronomer/providers/amazon/aws/sensors/s3.py [hunk 1]
```diff
@@ -84,5 +90,226 @@ def execute(self, context: Dict) -> Any:
...
+        self, context: Dict[Any, Any], event: Any = None
+    ) -> None:  # pylint: disable=unused-argument
+        if event["status"] == "error":
+            raise AirflowException(event["message"])
+        return None
+
+
+class S3PrefixSensorAsync(BaseOperator):
+    """
+    Waits for a prefix or all prefixes to exist. A prefix is the first part of a key,
```
Comment (sunank200) [inline astronomer/providers/amazon/aws/sensors/s3.py:253]: The doc string might need comment with respect to async implementation.

File: astronomer/providers/amazon/aws/sensors/s3.py [hunk 2]
```diff
@@ -84,5 +90,226 @@ def execute(self, context: Dict) -> Any:
...
+        self,
+        *,
+        bucket_name: str,
+        prefix: Union[str, List[str]],
+        delimiter: str = '/',
+        aws_conn_id: str = 'aws_default',
+        verify: Optional[Union[str, bool]] = None,
+        **kwargs: Any,
+    ):
+        super().__init__(**kwargs)
```
Comment (sunank200) [inline astronomer/providers/amazon/aws/sensors/s3.py:289]: `super().__init__(**kwargs)` might not be required here as we are inheriting from `BaseOperator`

File: astronomer/providers/amazon/aws/triggers/s3.py [hunk 3]
```diff
@@ -18,95 +16,249 @@ def __init__(
...
-                self.log.info("Poking for key : s3://%s/%s", self.bucket_name, self.bucket_key)
-                if await self._check_key(client, self.bucket_name, self.bucket_key, self.wildcard_match):
-                    yield TriggerEvent(True)
+        try:
+            async with await hook.get_client_async() as client:
+                while True:
+                    self.log.info('Poking for prefix : %s in bucket s3://%s', self.prefix, self.bucket_name)
+                    buffer = []
+                    for elem in (
+                        hook._check_for_prefix(client, prefix, self.delimiter, self.bucket_name)
```
Comment (sunank200) [inline astronomer/providers/amazon/aws/triggers/s3.py:252]: Isn't this `hook._check_for_prefix(client, prefix, self.delimiter, self.bucket_name)` returning `bool`? Is this for loop iterating over the each prefix?

File: tests/amazon/aws/hooks/test_s3_hooks.py [hunk 4]
```diff
@@ -35,3 +41,368 @@ async def test_aws_base_hook_async_get_client_async_with_aws_secrets(mock_get_co
...
+@mock.patch("astronomer.providers.amazon.aws.hooks.s3.S3HookAsync.list_prefixes")
+@pytest.mark.asyncio
+async def test_s3_prefix_sensor_hook_check_for_prefix(mock_list_prefixes, mock_client):
+    """
+    Test is_key_unchanged gives AirflowException
+    :return:
+    """
+    mock_list_prefixes.return_value = ['async-prefix1/', 'async-prefix2/']
+
+    s3_hook_async = S3HookAsync(client_type="S3", resource_type="S3")
```
Comment (sunank200) [inline tests/amazon/aws/hooks/test_s3_hooks.py:387]: These three cases can be parameterised for the tests.

**Review by sunank200 — COMMENTED** [review]
Overall, LGTM. I have added few minor comments and a question.

File: astronomer/providers/amazon/aws/triggers/s3.py (hunk 3 above)
Comment (phanikumv) [reply astronomer/providers/amazon/aws/triggers/s3.py:252]: it returns a Coroutine[bool]

**Review by bharanidharan14 — COMMENTED** [review]
Overall looks fine for me, only concerns is `any` type in `__init__`

File: astronomer/providers/amazon/aws/sensors/s3.py (hunk 2 above)
Comment (phanikumv) [reply astronomer/providers/amazon/aws/sensors/s3.py:289]: it is required other wise params like task_id wont be recognized. Getting DAG error if removed.

File: tests/amazon/aws/hooks/test_s3_hooks.py (hunk 4 above)
Comment (phanikumv) [reply tests/amazon/aws/hooks/test_s3_hooks.py:402]: this is completed

**Comment by kaxil** [comment]
Is this still work in progress? or ready for final review?

File: astronomer/providers/amazon/aws/example_dags/example_s3.py [hunk 5]
```diff
@@ -1,15 +1,28 @@
-from datetime import timedelta
+import os
+from datetime import datetime, timedelta
 
-from airflow.models.dag import DAG
-from airflow.utils.timezone import datetime
+from airflow import DAG
```
Comment (kaxil) [inline astronomer/providers/amazon/aws/example_dags/example_s3.py:4]: Why this change?

File: astronomer/providers/amazon/aws/example_dags/example_s3.py [hunk 6]
```diff
@@ -1,15 +1,28 @@
...
 }
 
+PROJECT_ID = os.environ.get("GCP_PROJECT_ID", "astronomer-airflow-providers")
+S3_BUCKET_NAME = os.environ.get("S3_BUCKET_NAME", "test-team-providers")
+S3_BUCKET_KEY = os.environ.get("S3_BUCKET_KEY", "sample_key.txt")
+S3_BUCKET_WILDCARD_KEY = os.environ.get("S3_BUCKET_WILDCARD_KEY", "test*")
+WILDCARD_KEY_FALSE = os.environ.get("WILDCARD_KEY_FALSE", False)
+WILDCARD_KEY_TRUE = os.environ.get("WILDCARD_KEY_TRUE", True)
+PREFIX = os.environ.get("PREFIX", "test")
+INACTIVITY_PERIOD = os.environ.get("INACTIVITY_PERIOD", 5)
```
Comment (kaxil) [inline astronomer/providers/amazon/aws/example_dags/example_s3.py:24]:
```suggestion
INACTIVITY_PERIOD = float(os.environ.get("INACTIVITY_PERIOD", 5))
```

File: astronomer/providers/amazon/aws/example_dags/example_s3.py [hunk 7]
```diff
@@ -1,15 +1,28 @@
...
     "retry": 5,
     "retry_delay": timedelta(minutes=1),
 }
 
+PROJECT_ID = os.environ.get("GCP_PROJECT_ID", "astronomer-airflow-providers")
+S3_BUCKET_NAME = os.environ.get("S3_BUCKET_NAME", "test-team-providers")
+S3_BUCKET_KEY = os.environ.get("S3_BUCKET_KEY", "sample_key.txt")
+S3_BUCKET_WILDCARD_KEY = os.environ.get("S3_BUCKET_WILDCARD_KEY", "test*")
+WILDCARD_KEY_FALSE = os.environ.get("WILDCARD_KEY_FALSE", False)
+WILDCARD_KEY_TRUE = os.environ.get("WILDCARD_KEY_TRUE", True)
```
Comment (kaxil) [inline astronomer/providers/amazon/aws/example_dags/example_s3.py:22]:
```suggestion
WILDCARD_KEY_FALSE = bool(os.environ.get("WILDCARD_KEY_FALSE", False))
WILDCARD_KEY_TRUE = bool(os.environ.get("WILDCARD_KEY_TRUE", True))
```

as `os.environ.get` will always return a String or None

**Comment by kaxil** [comment]
There were a lot of conflicts here, I suggest [using rebase](http://stackoverflow.com/a/7244456/1110993) instead of merge commits

File: astronomer/providers/amazon/aws/example_dags/example_s3.py (hunk 5 above)
Comment (phanikumv) [reply astronomer/providers/amazon/aws/example_dags/example_s3.py]: Fixed post the rebase from main

File: astronomer/providers/amazon/aws/example_dags/example_s3.py (hunk 6 above)
Comment (phanikumv) [reply astronomer/providers/amazon/aws/example_dags/example_s3.py]: Fixed post the rebase from main

File: astronomer/providers/amazon/aws/sensors/s3.py [hunk 8]
```diff
@@ -68,6 +67,7 @@ def __init__(
 
     def _resolve_bucket_and_key(self) -> None:
         """If key is URI, parse bucket"""
+        print("In _resolve_bucket_and_key")
```
Comment (sunank200) [inline astronomer/providers/amazon/aws/sensors/s3.py:70]: Print statement can be removed.

**Review by sunank200 — COMMENTED** [review]
Overall LGTM. Only two minor comments to remove the `print` statements

File: astronomer/providers/amazon/aws/sensors/s3.py [hunk 9]
```diff
@@ -58,7 +57,7 @@ def __init__(
         **kwargs: Any,
     ):
         super().__init__(**kwargs)
-
+        print("In init of S3KeySensorAsync")
```
[the remark this answers is not in the document]
Comment (phanikumv) [reply astronomer/providers/amazon/aws/sensors/s3.py]: removed this print statement

File: astronomer/providers/amazon/aws/sensors/s3.py (hunk 8 above)
Comment (phanikumv) [reply astronomer/providers/amazon/aws/sensors/s3.py]: removed this print statement

**Review by sunank200 — APPROVED** [review]

File: astronomer/providers/amazon/aws/hooks/s3.py [hunk 10]
```diff
@@ -45,6 +45,52 @@ async def _check_exact_key(client: AioBaseClient, bucket: str, key: str) -> bool
...
+        :param prefix: a key prefix
+        :type prefix: str
+        :param delimiter: the delimiter marks key hierarchy.
+        :type delimiter: str
+        :param page_size: pagination size
+        :type page_size: int
+        :param max_items: maximum items to return
+        :type max_items: int
+        :return: a list of matched prefixes
+        :rtype: list
```
Comment (kaxil) [inline astronomer/providers/amazon/aws/hooks/s3.py:72]:
```suggestion
        :param client: ClientCreatorContext
        :param bucket_name: the name of the bucket
        :param prefix: a key prefix
        :param delimiter: the delimiter marks key hierarchy.
        :param page_size: pagination size
        :param max_items: maximum items to return
        :return: a list of matched prefixes
```

We don't need type since we have Typehints :)

File: astronomer/providers/amazon/aws/hooks/s3.py [hunk 11]
```diff
@@ -81,6 +127,34 @@ async def check_key(self, client: AioBaseClient, bucket: str, key: str, wildcard
...
+        Checks that a prefix exists in a bucket
+
+        :param bucket_name: the name of the bucket
+        :type bucket_name: str
+        :param prefix: a key prefix
+        :type prefix: str
+        :param delimiter: the delimiter marks key hierarchy.
+        :type delimiter: str
+        :return: False if the prefix does not exist in the bucket and True if it does.
+        :rtype: bool
```
Comment (kaxil) [inline astronomer/providers/amazon/aws/hooks/s3.py:150]:
```suggestion
        :param bucket_name: the name of the bucket
        :param prefix: a key prefix
        :param delimiter: the delimiter marks key hierarchy.
        :return: False if the prefix does not exist in the bucket and True if it does.
```

File: astronomer/providers/amazon/aws/triggers/s3.py [hunk 12]
```diff
@@ -201,3 +201,63 @@ async def run(self) -> AsyncIterator["TriggerEvent"]:  # type: ignore[override]
 
     def _get_async_hook(self) -> S3HookAsync:
         return S3HookAsync(aws_conn_id=self.aws_conn_id, verify=self.verify)
+
+
+class S3PrefixSensorTrigger(BaseTrigger):
```
Comment (kaxil) [inline astronomer/providers/amazon/aws/triggers/s3.py:206]:
```suggestion
class S3PrefixTrigger(BaseTrigger):
```

File: astronomer/providers/amazon/aws/hooks/s3.py (hunk 11 above)
Comment (phanikumv) [reply astronomer/providers/amazon/aws/hooks/s3.py]: completed these changes

File: astronomer/providers/amazon/aws/triggers/s3.py (hunk 12 above)
Comment (phanikumv) [reply astronomer/providers/amazon/aws/triggers/s3.py]: completed these changes

File: astronomer/providers/amazon/aws/example_dags/example_s3.py (hunk 7 above)
Comment (phanikumv) [reply astronomer/providers/amazon/aws/example_dags/example_s3.py]: completed these changes

**Review by kaxil — APPROVED** [review]
